[PROMOTION][RULE]Add comparison operator for Cart Quantity and Item Total rules

Prometee opened the pull request:
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | none
| License         | MIT

Add a comparison operator to be able to make more complex rules instead of having only >= comparison operator available.
This PR allow to choose among basic comparison operators : ===, !==, <, >, <= or >= (the default legacy one)

![image](https://user-images.githubusercontent.com/861820/106498736-f8c31e00-64bf-11eb-8fbc-3d7e2899ee2a.png)

Prometee committed:
Add comparison operator to CartQuantityRuleChecker and ItemTotalRuleChecker

Prometee committed:
Add comparison operator to CartQuantityRuleChecker and ItemTotalRuleChecker

# Conflicts:
#	src/Sylius/Component/Promotion/Checker/Rule/ItemTotalRuleChecker.php

Prometee committed:
Add factory enhancement

Prometee committed:
Add factory enhancement

Prometee committed:
Fix php spec tests

AdamKasp commented on src/Sylius/Bundle/PromotionBundle/Resources/translations/messages.en.yml:
```diff
@@ -50,12 +50,30 @@ sylius:
                 variants: Products
             item_total_configuration:
                 amount: Amount
+                comparison_operator:
+                    label: Comparison operator
+                    choices:
+                        greater_than_equal: Greater than or equal (>=)
+                        equal: Equal (===)
+                        different: Different (!==)
+                        lower_than: Lower than (<)
+                        lower_than_equal: Lower than or equal (<=)
+                        greater_than: Greater than (>)
```
Do we need this duplication?
I think that a common translation key for both item and cart will be enough.

AdamKasp commented:
And Behats are appreciated. :)

Prometee commented:
Do not stale

lchrusciel reviewed (commented):
Hey Francis,

thanks a lot for contribution. I've requested @CoderMaggie to decide if we want to have it in the codebase from the feature side.

From my side, we cannot merge this PR at this state for following reasons:
* Lack of behat tests (to cover these functionalities from the business perspective) or phpunit test (to cover these functionalities from the operating perspective)
* Lack of specs 
* Code duplication - the choice type could be extracted to separate class. In addition comparison logic could be extracted as well - just for sack of simplicity of testing each case
But before you will do any of that, lets see if we want to have it in our codebase

CoderMaggie commented:
it took some time for me to get there, sorry 😓 

Of course, we do want to have this feature in Core 🙏🏻 . @Prometee will you be able to continue working on this? We're awaiting tests as @lchrusciel mentioned above.

Prometee commented:
@CoderMaggie Yes sure, I will make some refactoring as @lchrusciel suggested.

Prometee committed:
[Promotion][Rule] Extract comparison operator matching to dedicated service

- Add ComparisonOperatorMatcher service + interface in Promotion component
- Reuse matcher in CartQuantity and ItemTotal rule checkers
- Reuse matcher-provided operators in promotion rule form types
- Wire matcher into DI config for forms and checkers
- Add/adjust PHPUnit coverage for matcher and affected rule checkers

Prometee committed:
[Promotion] Preserve rule checker constructor BC

Prometee committed:
[Promotion] Fix comparison matcher import

GSadee commented on src/Sylius/Bundle/PromotionBundle/Form/Type/Rule/CartQuantityConfigurationType.php:
```diff
@@ -13,23 +13,52 @@
 
 namespace Sylius\Bundle\PromotionBundle\Form\Type\Rule;
 
+use Sylius\Component\Promotion\Checker\Comparison\ComparisonOperatorMatcher;
+use Sylius\Component\Promotion\Checker\Comparison\ComparisonOperatorMatcherInterface;
 use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
 use Symfony\Component\Form\Extension\Core\Type\IntegerType;
 use Symfony\Component\Form\FormBuilderInterface;
 
 final class CartQuantityConfigurationType extends AbstractType
 {
+    public function __construct(private ?ComparisonOperatorMatcherInterface $comparisonOperatorMatcher = null)
+    {
+    }
```
Deprecate not passing this matcher and note it in the UPGRADE file, the same comment applies to the other form type and checkers

GSadee commented on src/Sylius/Component/Core/Factory/PromotionRuleFactoryInterface.php:
```diff
@@ -23,9 +23,9 @@
+    public function createCartQuantity(int $count, string $comparisonOperator = '>='): PromotionRuleInterface;
 
-    public function createItemTotal(string $channelCode, int $amount): PromotionRuleInterface;
+    public function createItemTotal(string $channelCode, int $amount, string $comparisonOperator = '>='): PromotionRuleInterface;
```
There are two ways to resolve this: either actually extend these methods in the interface, but document this correctly in the UPGRADE file, although this may still be considered a BC break, or do not extend these methods in the interface, but instead add additional optional arguments to the classes + document that these methods will change in 3.0.

GSadee replied:
What is more, the default comparison operator could be extracted to const

Prometee committed:
Deprecate optional comparison operator arguments in PromotionRuleFactory methods

# Conflicts:
#	UPGRADE-2.3.md

Prometee committed:
[Promotion] Trigger deprecation warning for missing ComparisonOperatorMatcher in rule checkers

Prometee committed:
Add ComparisonOperatorExists validation and related constraints

Prometee committed:
Deprecate optional comparison operator arguments in PromotionRuleFactory methods

Prometee committed:
[Promotion] Enhance validation for comparison operators in rule configurations

# Conflicts:
#	tests/Api/Responses/admin/promotion/post_promotion_with_invalid_rules_response.json

GSadee commented on src/Sylius/Bundle/PromotionBundle/Validator/Constraints/ComparisonOperatorExists.php:
```diff
@@ -0,0 +1,57 @@
+#[\Attribute]
+final class ComparisonOperatorExists extends Constraint
+{
+    /**
+     * @param array<string, mixed>|null $options
+     */
+    #[HasNamedArguments]
+    public function __construct(
+        ?array $options = null,
+        public string $message = 'sylius.promotion_rule.comparison_operator.invalid',
+        ?array $groups = null,
+        mixed $payload = null,
+    ) {
+        if (\is_array($options)) {
+            trigger_deprecation(
+                'sylius/promotion-bundle',
+                '2.3',
+                'Passing an array of options to configure the "%s" constraint is deprecated and will be removed in Sylius 3.0, use named arguments instead.',
+                static::class,
+            );
+
+            $this->message = $options['message'] ?? $this->message;
+            $groups ??= $options['groups'] ?? null;
+            $payload ??= $options['payload'] ?? null;
+        }
+
+        parent::__construct(groups: $groups, payload: $payload);
+    }
```
As this is newly created constraint, there is no need to maintain options and trigger deprecation here

```suggestion
    #[HasNamedArguments]
    public function __construct(
        public string $message = 'sylius.promotion_rule.comparison_operator.invalid',
        ?array $groups = null,
        mixed $payload = null,
    ) {
        parent::__construct(groups: $groups, payload: $payload);
    }
```
